Implement `willDisplay` and `didEndDisplaying` APIs

nuomi1 opened the pull request:
Closes #96

Implements `willDisplay` and `didEndDisplaying`.

nuomi1 commented:
I found that `collectionView(_:didEndDisplaying:forItemAt:)` may trigger a crash, but I don’t know how to fix it, should we use guard to check `indexPath`?

![CleanShot 2024-07-28 at 18 58 33@2x](https://github.com/user-attachments/assets/ca9d664e-c977-461d-a237-ae771ea6382c)

nuomi1 commented:
We can use the code from https://github.com/jessesquires/ReactiveCollectionsKit/pull/122 to reproduce the crash.

nuomi1 commented:
These are crash logs:

```text
critical 2024-07-29 15:11:40 +0000 numberOfSections(in:) ObjectIdentifier(0x00007b7c00007000) 3
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 0 5
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 1 1
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 2 13
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [0, 0]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [0, 1]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [0, 2]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [0, 3]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [0, 4]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [1, 0]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 0]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 1]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 2]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 3]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 4]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 5]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 6]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 7]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 8]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 9]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 10]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 11]
critical 2024-07-29 15:11:40 +0000 collectionView(_:willDisplay:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 12]
critical 2024-07-29 15:11:40 +0000 numberOfSections(in:) ObjectIdentifier(0x00007b7c00007000) 2
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 0 5
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 1 1
critical 2024-07-29 15:11:40 +0000 numberOfSections(in:) ObjectIdentifier(0x00007b7c00007000) 2
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 0 5
critical 2024-07-29 15:11:40 +0000 collectionView(_:numberOfItemsInSection:) ObjectIdentifier(0x00007b7c00007000) 1 1
critical 2024-07-29 15:11:41 +0000 collectionView(_:didEndDisplaying:forItemAt:) ObjectIdentifier(0x00007b7c00007000) [2, 0]
```

jessesquires commented:
@nuomi1 hm... let me see if I can repro and figure out what's wrong

jessesquires commented:
> These are crash logs:

Ahh. I think I see what's happening. 

1. we update the view model and apply the diff
2. the collection view will remove the cells in the view layer and the `didEndDisplaying` delegate methods will be called
3. however, we have already replaced the view model so now those index paths are invalid
4. crash 💥 

I played around with the Example App on this branch and it looks like the `willDisplay` / `didEndDisplaying` methods **are working correctly** -- so that's good. However, I did reproduce the crash when deleting all items.

These methods are primarily used for scrolling, when views come in and out of view. But they are also called when items are on-screen and get deleted. In this later case, that means we'll "drop" these events. However, I think that's probably ok. Folks should not be relying on these methods for specific state information.

jessesquires reviewed (changes requested):
Ok @nuomi1, I pushed some fixes to this branch and I think it's all good now!

Thanks for doing this! 😄 💯 

One last thing: could you [add a Changelog](https://github.com/jessesquires/ReactiveCollectionsKit/blob/main/CHANGELOG.md) entry for this? 

- Let's add a new version in the changelog, 0.1.4
- Follow the existing pattern for entries

nuomi1 commented:
> Ok @nuomi1, I pushed some fixes to this branch and I think it's all good now!
> 
> Thanks for doing this! 😄 💯
> 
> One last thing: could you [add a Changelog](https://github.com/jessesquires/ReactiveCollectionsKit/blob/main/CHANGELOG.md) entry for this?
> 
> * Let's add a new version in the changelog, 0.1.4
> * Follow the existing pattern for entries

Done.

nuomi1 commented:
I added some tests related to `CollectionViewDriver` and fixed the bug in `_safeSupplementaryViewModel`. If this PR is too big, we can split it into two.

jessesquires commented:
> I added some tests related to `CollectionViewDriver` and fixed the bug in `_safeSupplementaryViewModel`. If this PR is too big, we can split it into two.

thanks @nuomi1!
